### ISSUE TITLE
EarlyStopping should restore weights on end of training, not end of epoch

### ISSUE THREAD

**Issue by Stigjb** [tensorflow/tensorflow#35634]
## System information
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Darwin-19.2.0-x86_64-i386-64bit
- TensorFlow installed from (source or binary): Pipenv
- TensorFlow version (use command below): 2.0.0
- Python version: 3.7.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

## Current behavior

The EarlyStopping callback will restore the best weights only if it itself stops training. For example, if a model is set to train for 40 epochs, uses early stopping with a patience of 5, and the best weights occur in epoch 37, then the best weights will not be restored, and the model will have the weights from epoch 40.

## Expected behavior

EarlyStopping should, when initialized with `restore_best_weights=True`, always restore the best weights when training stops, regardless of the reason why training stopped.

## Code to reproduce the issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import tensorflow as tf
from numpy.random import RandomState
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential

N_CLASSES, N_SAMPLES = 5, 500

for seed in [1, 2]:
    print(f"=== Random seed is {seed} ===")

    tf.random.set_seed(seed)
    rng = RandomState(seed)

    x_train = rng.standard_normal(size=(N_SAMPLES, 10))
    x_test = rng.standard_normal(size=(N_SAMPLES, 10))
    y_train = rng.random_integers(N_CLASSES, size=N_SAMPLES)
    y_test = rng.random_integers(N_CLASSES, size=N_SAMPLES)

    model = Sequential([Dense(32), Dense(N_CLASSES)])
    model.compile("adam", "categorical_crossentropy", ["accuracy"])
    early_stopping = EarlyStopping(
        "val_accuracy", patience=10, verbose=1, restore_best_weights=True
    )
    history = model.fit(
        x_train,
        y_train,
        epochs=20,
        callbacks=[early_stopping],
        verbose=0,
        validation_data=(x_test, y_test),
    )
    best_acc = max(history.history["val_accuracy"])
    _, eval_acc = model.evaluate(x_test, y_test, verbose=0)
    print(f"Best accuracy in training: {best_acc}. In evaluation: {eval_acc}\n")
```

### Output from example:

```txt
=== Random seed is 1 ===
Restoring model weights from the end of the best epoch.
Epoch 00011: early stopping
Best accuracy in training: 0.19200000166893005. In evaluation: 0.19200000166893005

=== Random seed is 2 ===
Best accuracy in training: 0.15199999511241913. In evaluation: 0.14000000059604645

```

## Other info / logs

- Numpy version: 1.18.1

**Comment by Stigjb** [comment]
See this issue in keras-team/keras for several people confirming the issue: https://github.com/keras-team/keras/issues/12511

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the [Gist](https://colab.sandbox.google.com/gist/amahendrakar/3143abc3aeaa7962f38a97097464c5ed/35634.ipynb) here. Thanks!

**Comment by morelen17** [comment]
@Stigjb current behavior isn't clear at all, I agree with you. But it is not wrong. Just consider a situation: you train your model for 20 epochs, EarlyStopping callback isn't triggered and you want to continue the training process. If your *Expected behavior* was applied and the best weights were restored, then the training would continue from e.g. 15th epoch, but not from 20th as you would expect.

So I believe that it isn't the bug, but it definitely should be documented somewhere, e.g. in the description of `restore_best_weights` argument.

**Comment by mcmar** [comment]
I think the best solution to this is to also add a `restore_best_weights` option to `Model.fit` that you can have this behavior on early termination and regular termination

**Comment by iliaschalkidis** [comment]
> I think the best solution to this is to also add a `restore_best_weights` option to `Model.fit` that you can have this behavior on early termination and regular termination

In a similar fashion, we just hijacked the original callback and enforced the callback to replace the weights from the "best" epoch in the end of the training process.

```python
from tensorflow.keras.callbacks import EarlyStopping

class ReturnBestEarlyStopping(EarlyStopping):
    def __init__(self, **kwargs):
        super(ReturnBestEarlyStopping, self).__init__(**kwargs)

    def on_train_end(self, logs=None):
        if self.stopped_epoch > 0:
            if self.verbose > 0:
                print(f'\nEpoch {self.stopped_epoch + 1}: early stopping')
        elif self.restore_best_weights:
            if self.verbose > 0:
                print('Restoring model weights from the end of the best epoch.')
            self.model.set_weights(self.best_weights
```

**Comment by darrylmelander** [comment]
> @Stigjb current behavior isn't clear at all, I agree with you. But it is not wrong. Just consider a situation...<discussion of continued training>...I believe that it isn't a bug, but it definitely should be documented somewhere...

Failing to restore the weights is not the behavior of least surprise, which by some definitions makes it a bug no matter what the documentation says.  At best, it is ill-advised.

I'm not fond of the idea of putting it into Model.fit either.  I think it does belong in EarlyStopping because that class already encapsulates the concepts of tracking something other than "normal" weights and metrics and making choices based on it.  

In my mind, the _only_ correct default behavior is to always restore the best weights when restore_best_weights is True.  If we want to make it possible to override this sensible default behavior for the sake of resuming training, we should add yet another flag to the EarlyStopping constructor.  Anything else leads to behavior most people won't expect.  Don't force people into documentation archeology, digging around until they discover a surprise.

**Comment by k1moradi** [comment]
I'm wondering in "on_epoch_end" function should not 
`self.best_weights = self.model.get_weights()`
be like
`self.best_weights = self.model.get_weights().copy()`
?

**Comment by iliaschalkidis** [comment]
> I'm wondering in "on_epoch_end" function should not
> `self.best_weights = self.model.get_weights()`
> be like
> `self.best_weights = self.model.get_weights().copy()`
> ?

You could... But why? 😃

**Comment by k1moradi** [comment]
> You could... But why? 😃

I'm not sure how get_weights() function works. Imagine if it returns only a pointer to the weights, then what happens if model is trained more? It might be able to change the weights. No?

**Comment by iliaschalkidis** [comment]
> > You could... But why? 😃
> 
> I'm not sure how get_weights() function works. Imagine if it returns only a pointer to the weights, then what happens if model is trained more? It might be able to change the weights. No?

Ok, I get it.  Maybe it's an extra precaution, but if this the case then the original, quite ill-adviced, implementation of `EarlyStopping` is also wrong / buggy: https://github.com/tensorflow/tensorflow/blob/e5cc10bcb5140ef8478a58c2ab260c5dcb2b1315/tensorflow/python/keras/callbacks.py#L1742

I doubt that such a huge bug resides there for almost 3 years.....

This seems to be the actual implementation of `get_weights()`:   https://github.com/tensorflow/tensorflow/blob/e5cc10bcb5140ef8478a58c2ab260c5dcb2b1315/tensorflow/python/keras/engine/base_layer.py#L1853

**Comment by k1moradi** [comment]
> I doubt that such a huge bug resides there for almost 3 years.....

You are right. Get_weight function does not need copy(). Interestingly I have reasons to believe that the "restore best weights" is not working for me, but I have already checked and it seems using get_weight().copy() is not the solution. 

During training, the running values of a loss function and another metric are getting saved in training logs. We use those values to stop training. We can also evaluate the model after training ends and get the best values coming from the best model. I expect the minimum value in logs matches the final restored model evaluation values, but it never happens in my case. "model.evaluate" function reports the last epochs training results, not the best model. I made a custom early_stopping function that stops training when loss value reaches a certain goal. Still, the "model.evaluate" output is different from the best values stored in the training logs, which is odd. Tensorflow is powerful but still has a lot of bugs in my limited experience.

For example, in the following plot, I was monitoring mean absolute percentage error (MAPE) to stop training and patience was 10000 epochs. At epoch 39294 training stoped and early stopping supposed to restore the best weights. Model.evaluate reports a MAPE of 103% while the best MAPE is ~85% and happening at epoch 29294.
![image](https://user-images.githubusercontent.com/18602635/92831230-250cbf80-f3a4-11ea-978a-45a1106004d1.png)

In the following plot, I used my own custom early stopping callback that stops training immediately after MAPE becomes smaller than 90%. So there is no patience involved. Still, MAPE values I get from model.evaluate after training is different from the minimum MAPE I find in logs, and the loss in model accuracy is not negligible.
![image](https://user-images.githubusercontent.com/18602635/92832066-2e4a5c00-f3a5-11ea-848e-668d7de06b8c.png)

**Comment by sachinprasadhs** [comment]
@Stigjb , Tried to reproduce your issue but got shape error, please find the gist [here].(https://colab.research.google.com/gist/sachinprasadhs/c0630d82e37522b119a7764cfb49be7d/35634.ipynb)

**Comment by durgals** [comment]
At the beginning, I have issue of the shape errors. So I had to apply sklearn.preprocessing.OneHotEncoder. 
But with tensorflow version 2.9.1, I do not have that issue. if restore_best_weights = True, it always loads the best weights. 

![image](https://user-images.githubusercontent.com/10932215/185300498-f8fcb321-5299-4270-9c5b-fa815d00839c.png)

**Comment by Red54** [comment]
> At the beginning, I have issue of the shape errors. So I had to apply sklearn.preprocessing.OneHotEncoder. But with tensorflow version 2.9.1, I do not have that issue. if restore_best_weights = True, it always loads the best weights.
> 
> ![image](https://user-images.githubusercontent.com/10932215/185300498-f8fcb321-5299-4270-9c5b-fa815d00839c.png)

Because it triggers EarlyStopping:
```
Epoch 17: early stopping
```
